phoenix-core-client/src/main/java/org/apache/phoenix/util/json/BsonDataFormat.java

```diff
@@ -0,0 +1,201 @@
+        return Bytes.toBytes(((RawBsonDocument) object).getByteBuffer().asNIO());
+    }
+
+    @Override
+    public Object toObject(String value) {
+        return RawBsonDocument.parse(value);
+    }
+
+    @Override
+    public Object toObject(byte[] bytes, int offset, int length) {
+        return new RawBsonDocument(bytes, offset, length);
+    }
+
+    @Override
+    public int estimateByteSize(Object o) {
+        RawBsonDocument rawBSON = (RawBsonDocument) o;
+        return rawBSON.size();
+    }
+
+    @Override
+    public int getValueType(Object obj, String jsonPathExprStr) {
+        BsonValue value = getBsonValue(jsonPathExprStr, (RawBsonDocument) obj);
+        return getSqlType(value);
+    }
+
+    @Override
+    public Object getValue(Object obj, String jsonPathExprStr) {
+        BsonValue value = getBsonValue(jsonPathExprStr, (RawBsonDocument) obj);
+        return getValue(value);
+    }
+
+    private Object getValue(BsonValue value) {
+        if (value != null) {
+            switch (value.getBsonType()) {
+            case INT32:
+                return value.asInt32().getValue();
+            case INT64:
+                return value.asInt64().getValue();
+            case STRING:
+            case SYMBOL:
+                return value.asString().getValue();
+            case DECIMAL128:
+                return value.asDecimal128().doubleValue();
+            case DOUBLE:
+                return value.asDouble().getValue();
+            case BOOLEAN:
+                return value.asBoolean().getValue();
+            case BINARY:
+                return value.asBinary().getData();
+            case DATE_TIME:
+                return value.asDateTime().getValue();
+            case DOCUMENT:
+                return value.asDocument().toJson();
+            case ARRAY:
+                return readArray(value).toString();
+            default:
+                return null;
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public ByteBuffer updateValue(Object top, String jsonPathExprStr, String newVal) {
+        Configuration conf = Configuration.builder().jsonProvider(new BsonJsonProvider()).build();
+        BsonValue newValue = JsonPath.using(conf).parse(newVal).json();
+        BsonDocument root = fromRaw((RawBsonDocument) top);
+        JsonPath.using(conf).parse(root).set(jsonPathExprStr, newValue);
+        RawBsonDocument
+                updated =
+                new RawBsonDocumentCodec().decode(new BsonDocumentReader(root),
+                        DecoderContext.builder().build());
+        return updated.getByteBuffer().asNIO();
+    }
+
+    // Ref: https://github.com/json-path/JsonPath/pull/828
+    @Override
+    public boolean isPathValid(Object top, String path) {
+        try{
+            Configuration conf = Configuration.builder().jsonProvider(new BsonJsonProvider()).build();
+            BsonDocument root = fromRaw((RawBsonDocument) top);
```

Comment:
Same here, can we just type cast `BsonDocument root = ((RawBsonDocument) top).asDocument()` ?

Reply:
Will take this another iteration with additional feature